File: strategy_pattern/ruby/mallard_duck.rb
# frozen_string_literal: true

require_relative 'duck'
# A specific type of normal flying quacking duck
class MallardDuck < Duck
  def initialize
    super(QuackBehavior::Quack, FlyBehavior::FlyWithWings)
  end

  def display
    puts "I'm a mallard duck"
  end
end

File: c_plus_plus/strategy_pattern/MallardDuck.cpp
//
// Created by Randy Layne on 5/10/20.
//

#include "MallardDuck.h"
#include "FlyWithWings.h"
#include "Quacker.h"

MallardDuck::MallardDuck()
{
    // same as default duck, but want to make a point.
    // plus, if the default changes we are covered
    FlyBehavior *f = new FlyWithWings();
    QuackBehavior *q = new Quacker();

    setFlyBehavior(f);
    setQuackBehavior(q);
}


void MallardDuck::display(std::ostream &outs) const {
    outs << "I'm a real Mallard Duck.";
}
